uxarray/remap/inverse_distance_weighted.py

```diff
@@ -0,0 +1,269 @@
+        raise ValueError(
+            f"Invalid source_data shape. The final dimension should match the number of corner "
+            f"nodes ({source_grid.n_node}), edge nodes ({source_grid.n_edge}), or face centers ({source_grid.n_face}) "
+            f"in the source grid, but received: {source_data.shape}"
+        )
+
+    if coord_type == "spherical":
+        if remap_to == "nodes":
+            lon, lat = (
+                destination_grid.node_lon.values,
+                destination_grid.node_lat.values,
+            )
+        elif remap_to == "face centers":
+            lon, lat = (
+                destination_grid.face_lon.values,
+                destination_grid.face_lat.values,
+            )
+        elif remap_to == "edge centers":
+            lon, lat = (
+                destination_grid.edge_lon.values,
+                destination_grid.edge_lat.values,
+            )
+        else:
+            raise ValueError(
+                f"Invalid remap_to. Expected 'nodes', 'edge centers', or 'face centers', "
+                f"but received: {remap_to}"
+            )
+
+        _source_tree = source_grid.get_ball_tree(coordinates=source_data_mapping)
+
+        dest_coords = np.vstack([lon, lat]).T
+
+        distances, nearest_neighbor_indices = _source_tree.query(
+            dest_coords, k=k_neighbors
+        )
+
+    elif coord_type == "cartesian":
+        if remap_to == "nodes":
+            x, y, z = (
+                destination_grid.node_x.values,
+                destination_grid.node_y.values,
+                destination_grid.node_z.values,
+            )
+        elif remap_to == "face centers":
+            x, y, z = (
+                destination_grid.face_x.values,
+                destination_grid.face_y.values,
+                destination_grid.face_z.values,
+            )
+        elif remap_to == "edge centers":
+            x, y, z = (
+                destination_grid.edge_x.values,
+                destination_grid.edge_y.values,
+                destination_grid.edge_z.values,
+            )
+        else:
+            raise ValueError(
+                f"Invalid remap_to. Expected 'nodes', 'edge centers', or 'face centers', "
+                f"but received: {remap_to}"
+            )
+
+        _source_tree = source_grid.get_kd_tree(coordinates=source_data_mapping)
+
+        dest_coords = np.vstack([x, y, z]).T
+
+        distances, nearest_neighbor_indices = _source_tree.query(
+            dest_coords, k=k_neighbors
+        )
+
+    else:
+        raise ValueError(
+            f"Invalid coord_type. Expected either 'spherical' or 'cartesian', but received {coord_type}"
+        )
+
+    if nearest_neighbor_indices.ndim > 1:
+        nearest_neighbor_indices = nearest_neighbor_indices.squeeze()
+
+    weights = 1 / (distances**power + 1e-6)
+    weights /= np.sum(weights)
+
+    destination_data = np.sum(
+        source_data[..., nearest_neighbor_indices] * weights, axis=-1
+    )
```

Comment:
I'll give this a deep look tomorrow, but I presume the issue lies somewhere here in how we compute our weights.

Reply:
Oh, yeah that is a problem! I will start seeing if I can figure it out too, how did you make those plots? Could you send the code for that over Slack or in a gist or something? Having that for any remapping function would be great for testing.

Reply:
I think I found the problem. Each point gets its 8 nearest neighbors (by default), and calculates a weight based on those. The sum of those 8 should be equal to 1. The current way however takes sum of all the weights in the entire grid, instead of the 8 it should be for a single point, and divides the weights by that, thus making them extremely small, causing near zero values. I just added a loop in the weights calculation to properly make the weights get divided by the correct sum of weights. I will push my changes, could you let me know if this works on your plots?

Reply:
> I think I found the problem. Each point gets its 8 nearest neighbors (by default), and calculates a weight based on those. The sum of those 8 should be equal to 1. The current way however takes sum of all the weights in the entire grid, instead of the 8 it should be for a single point, and divides the weights by that, thus making them extremely small, causing near zero values. I just added a loop in the weights calculation to properly make the weights get divided by the correct sum of weights. I will push my changes, could you let me know if this works on your plots?

Let me know! I'll generate some more plots once you've got it resolved, but this definitely sounds like the issue.

Reply:
For these plots, I've been using the GeoDataFrame approach

```Python
%%time
downsampled_uxds_nn = uxds["refl10cm_max"].isel(Time=2).nearest_neighbor_remap(destination_obj = coarse_grid, remap_to="face centers")
downsampled_uxds_idw = uxds["refl10cm_max"].isel(Time=2).inverse_distance_weighted_remap(destination_obj = coarse_grid, remap_to="face centers")
lon_bounds = (-75, -71)
lat_bounds = (40, 42)
gdf_og = uxds["refl10cm_max"].isel(Time=2).subset.bounding_box(lon_bounds, lat_bounds).to_geodataframe()
gdf_nn = downsampled_uxds_nn.subset.bounding_box(lon_bounds, lat_bounds).to_geodataframe()
gdf_idw = downsampled_uxds_idw.subset.bounding_box(lon_bounds, lat_bounds).to_geodataframe()

(gdf_og.hvplot.polygons(rasterize=True, 
                              c='refl10cm_max', 
                              cmap=cmap, 
                              geo=True, 
                              width=1000, 
                              height=500,
                              grid=True, 
                              projection=projection,
                              title="Original Mesh") + \
    
    
    gdf_nn.hvplot.polygons(rasterize=True, 
                              c='refl10cm_max', 
                              cmap=cmap, 
                              geo=True, 
                              width=1000, 
                              height=500,
                              grid=True, 
                              projection=projection,
                              title = "Nearest Neighbor Remap") + \
 gdf_idw.hvplot.polygons(rasterize=True, 
                                 c='refl10cm_max',
                                 cmap=cmap, 
                                 geo=True, 
                                 width=1000, 
                                 height=500, 
                                 grid=True, 
                                 projection=projection,
                                 title = "IDW Remap")).cols(1).opts(fig_size=200)
```

Reply:
Awesome! Thanks for the code snippet. The changes should be up now, let me know what you find and if the changes create any new problems. Thanks for catching this error!

Reply:
![image](https://github.com/UXARRAY/uxarray/assets/67855069/bb6d6124-a071-419c-a91d-eb7bf2b4bb18)

This looks as expected!

Reply:
Once you address my other comments (the one about passing in more params through a dataset or dataarray), I'll generate more plots and do some additional testing.